File: js/src/_stream.ts
/**
 * NDJSON line reader (D3) — the hub's streaming responses are newline-delimited JSON, never actual
 * `text/event-stream`, matching the C# `IAsyncEnumerable<T>` and Python `Iterator`/`AsyncIterator`
 * shape. One line-buffering async generator, shared by `chatStream`/`generateStream`, decodes and
 * splits a `ReadableStream<Uint8Array>` on `\n`, buffering a partial trailing chunk across reads.
 *
 * Considered and rejected: `TextDecoderStream` + a `TransformStream` pipeline (D3) — more
 * "platform-idiomatic" on paper, but the manual buffering here is ~15 lines the corpus already
 * exercises, and it avoids a second code path for runtimes where `TextDecoderStream` lagged
 * `fetch`/`ReadableStream` support.
 */

/**
 * Decodes a byte stream into complete lines, one per yield. The final partial line (no trailing
 * `\n`) is flushed when the stream ends, exactly once, so a body with no trailing newline is not
 * silently dropped.
 */
export async function* readNdjsonLines(
  stream: ReadableStream<Uint8Array>,
): AsyncGenerator<string> {
  const reader = stream.getReader();
  const decoder = new TextDecoder("utf-8");
  let buffer = "";

  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) {
        break;
      }
      buffer += decoder.decode(value, { stream: true });
      let newlineIndex: number;
      while ((newlineIndex = buffer.indexOf("\n")) !== -1) {
        const line = buffer.slice(0, newlineIndex);
        buffer = buffer.slice(newlineIndex + 1);
        yield line;
      }
    }
    buffer += decoder.decode();
    if (buffer.length > 0) {
      yield buffer;
    }
  } finally {
    reader.releaseLock();
  }
}

/**
 * Groups an SSE line stream into `{event, data}` pairs on each blank-line frame boundary — the
 * mechanics phase-9's dotnet `SseFrameReader` and python's `parse_sse_lines` share between the
 * speech, image-job and admin streams. Built on {@link readNdjsonLines} (D4): the byte decoding
 * and line splitting is written once regardless of which line discipline — NDJSON or SSE — sits
 * on top of it. `data:` lines accumulate (multi-line payloads join with `\n`); a comment line
 * (`:`-prefixed) and any other field are ignored. A frame with no `data:` is skipped rather than
 * yielded as `{data: {}}`, since every InferHub SSE frame this client reads carries a JSON payload.
 */
export async function* readSseFrames(
  stream: ReadableStream<Uint8Array>,
): AsyncGenerator<{ event?: string; data: Record<string, unknown> }> {
  let event: string | undefined;
  let dataLines: string[] = [];

  const emit = (): { event?: string; data: Record<string, unknown> } | undefined => {
    if (dataLines.length === 0) return undefined;
    const raw = dataLines.join("\n");
    let data: Record<string, unknown>;
    try {
      data = JSON.parse(raw) as Record<string, unknown>;
    } catch {
      data = { raw };
    }
    return { event, data };
  };

  for await (const line of readNdjsonLines(stream)) {
    if (line === "") {
      const frame = emit();
      if (frame) yield frame;
      event = undefined;
      dataLines = [];
      continue;
    }
    if (line.startsWith(":")) continue;
    if (line.startsWith("event:")) {
      event = line.slice("event:".length).trim();
    } else if (line.startsWith("data:")) {
      dataLines.push(line.slice("data:".length).trim());
    }
  }
  const trailing = emit();
  if (trailing) yield trailing;
}

File: go/stream.go
package inferhub

import (
	"bufio"
	"io"
)

// NDJSON line reading (D2 in plans/phase-22-go-core.md) — the hub's streaming responses are
// newline-delimited JSON, never actual text/event-stream, matching the C# IAsyncEnumerable<T>,
// Python Iterator/AsyncIterator and TypeScript AsyncIterable<T> shape. bufio.Scanner already does
// exactly this buffering (a partial trailing line is held across reads, same as js's _stream.ts
// hand-rolled version); wrapping stdlib is the D5-budget-correct move, not reinventing it, and
// there is no test asserting the *technique* — only that a split line is not dropped.

// ChatStream iterates the chunks of a streaming POST /api/chat response, one ChatResponse per
// NDJSON line. Call Next until it returns false, then check Err. The zero value is not usable;
// obtain one from Client.ChatStream.
//
// D1 — this is a Scanner-shaped iterator (Next/Value/Err/Close), not a Go 1.23 iter.Seq2[T, error]
// range-over-func and not a channel. Three considered:
//   - iter.Seq2[ChatResponse, error]: the most "current" idiom, but it requires Go 1.23 and this
//     package cannot be build-tested in the environment phase 22 was implemented in (no local Go
//     toolchain — see plans/phase-22-go-core.md's verification section) — a generics/range-over-func
//     feature is exactly the kind of syntax an author gets subtly wrong without a compiler to check
//     it against, and go.mod pins 1.22 so the module works for a caller one version behind current.
//   - a channel + goroutine: needs a goroutine per stream and a caller who does not drain it to
//     completion leaks it; the Scanner shape has no concurrency to reason about at all.
//   - a callback (ForEach(func(ChatResponse) error) error): fine for "process every chunk" but
//     awkward for a caller who wants to break early after finding one thing, which `for
//     stream.Next()` does not require the caller to special-case.
//
// The Scanner shape is exactly database/sql.Rows and bufio.Scanner's own shape — the most-used
// iterator idiom in the stdlib itself, and every Go developer already knows the `for rows.Next()`
// loop.
type ChatStream struct {
	scanner   *bufio.Scanner
	closer    io.Closer
	servedBy  string
	sourceIDs []string
	cur       ChatResponse
	err       error
	done      bool
}

// Next advances to the next chunk. It returns false at end of stream, on a transport error, or
// after a terminal error chunk ({"error":...,"done":true}) — check Err to tell them apart.
func (s *ChatStream) Next() bool {
	if s.done {
		return false
	}
	for s.scanner.Scan() {
		line := s.scanner.Text()
		chunk, ok, err := parseNDJSONLine(line)
		if err != nil {
			s.err = err
			s.done = true
			return false
		}
		if !ok {
			continue // blank line
		}
		resp, err := chatResponseFromJSON(chunk)
		if err != nil {
			s.err = err
			s.done = true
			return false
		}
		resp.ServedBy = s.servedBy
		resp.SourceIDs = s.sourceIDs
		s.cur = resp
		if resp.Done != nil && *resp.Done {
			s.done = true
		}
		return true
	}
	if err := s.scanner.Err(); err != nil {
		s.err = err
	}
	s.done = true
	return false
}

// Value returns the chunk most recently produced by Next.
func (s *ChatStream) Value() ChatResponse { return s.cur }

// Err returns the first error encountered, if any — including a mid-stream terminal error chunk
// (root CLAUDE.md testing-discipline rule: a mid-stream error must terminate, never hang).
func (s *ChatStream) Err() error { return s.err }

// Close releases the underlying HTTP response body. Safe to call after Next has returned false,
// and safe to call more than once.
func (s *ChatStream) Close() error {
	if s.closer == nil {
		return nil
	}
	return s.closer.Close()
}

// GenerateStream is ChatStream's twin for POST /api/generate.
type GenerateStream struct {
	scanner   *bufio.Scanner
	closer    io.Closer
	servedBy  string
	sourceIDs []string
	cur       GenerateResponse
	err       error
	done      bool
}

func (s *GenerateStream) Next() bool {
	if s.done {
		return false
	}
	for s.scanner.Scan() {
		line := s.scanner.Text()
		chunk, ok, err := parseNDJSONLine(line)
		if err != nil {
			s.err = err
			s.done = true
			return false
		}
		if !ok {
			continue
		}
		resp, err := generateResponseFromJSON(chunk)
		if err != nil {
			s.err = err
			s.done = true
			return false
		}
		resp.ServedBy = s.servedBy
		resp.SourceIDs = s.sourceIDs
		s.cur = resp
		if resp.Done != nil && *resp.Done {
			s.done = true
		}
		return true
	}
	if err := s.scanner.Err(); err != nil {
		s.err = err
	}
	s.done = true
	return false
}

func (s *GenerateStream) Value() GenerateResponse { return s.cur }
func (s *GenerateStream) Err() error              { return s.err }
func (s *GenerateStream) Close() error {
	if s.closer == nil {
		return nil
	}
	return s.closer.Close()
}

// newScanner wraps a response body with a line-splitting bufio.Scanner sized generously above the
// default 64KiB token limit — a chat/generate chunk carrying a long tool-call payload could
// otherwise trip bufio.ErrTooLong, which the default js/python NDJSON readers (unbounded string
// buffering) never hit.
func newScanner(body io.Reader) *bufio.Scanner {
	scanner := bufio.NewScanner(body)
	buf := make([]byte, 0, 64*1024)
	scanner.Buffer(buf, 8*1024*1024)
	return scanner
}
